File: tests/core/contractors/codac2_tests_CtcVisible.cpp
/** * Codac tests - Visibility Contractors
 * ----------------------------------------------------------------------------
 * \date       2026
 * \author     Quentin Brateau
 * \copyright  Copyright 2026 Codac Team
 * \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_CtcVisible.h>
#include <codac2_Vector.h>

using namespace codac2;

TEST_CASE("CtcVisible & CtcNoVisible - Point/Segment visibility")
{
  // Observer at origin, obstacle horizontal from (2, -1) to (2, 1)
  Vector a({0.0, 0.0});
  Segment s({2.0, -1.0}, {2.0, 1.0});
  
  CtcVisible ctc_vis(a, s);
  CtcNoVisible ctc_no_vis(a, s);

  SECTION("Box fully in visible zone (in front of obstacle)")
  {
    IntervalVector x({{0.5, 1.5}, {-0.5, 0.5}});
    IntervalVector x_orig = x;

    ctc_vis.contract(x);
    CHECK(x == x_orig); // Should not be contracted

    ctc_no_vis.contract(x_orig);
    CHECK(x_orig.is_empty()); // Cannot be hidden if in front
  }

  SECTION("Box fully in hidden zone (shadow)")
  {
    IntervalVector x({{3.0, 4.0}, {0.1, 0.4}});
    IntervalVector x_orig = x;

    ctc_no_vis.contract(x_orig);
    CHECK(x == x_orig); // Fully hidden

    ctc_vis.contract(x);
    CHECK(x.is_empty()); // Fully in shadow -> not visible
  }

  SECTION("Box behind the observer")
  {
    IntervalVector x({{-2.0, -1.0}, {-1.0, -0.2}});
    IntervalVector x_orig = x;

    ctc_vis.contract(x);
    CHECK(x == x_orig); // Visible (obstacle is far away in the other direction)

    ctc_no_vis.contract(x_orig);
    CHECK(x_orig.is_empty());
  }

  SECTION("Box on the side (outside the angular cone)")
  {
    IntervalVector x({{1.0, 4.0}, {2.5, 4.0}}); // High above the obstacle
    IntervalVector x_orig = x;

    ctc_vis.contract(x);
    CHECK(x == x_orig); 

    ctc_no_vis.contract(x_orig);
    CHECK(x_orig.is_empty());
  }
}

File: tests/core/contractors/codac2_tests_CtcVisible.py
#!/usr/bin/env python

#  Codac tests - Visibility
# ----------------------------------------------------------------------------
#  \date       2026
#  \author     Quentin Brateau
#  \copyright  Copyright 2026 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestVisibility(unittest.TestCase):
  
  def test_CtcVisibility(self):
    # Observer at origin, obstacle vertical at x=2 from y=-1 to y=1
    a = [0.0, 0.0]
    s = Segment([2.0, -1.0], [2.0, 1.0])
    
    ctc_vis = CtcVisible(a, s)
    ctc_nvis = CtcNoVisible(a, s)

    # 1. Box fully in visible zone (in front of obstacle)
    x = IntervalVector([[0.5, 1.5], [-0.5, 0.5]])
    x_orig = IntervalVector(x)
    ctc_vis.contract(x)
    self.assertTrue(x == x_orig)

    x_test_no = IntervalVector(x_orig)
    ctc_nvis.contract(x_test_no)
    self.assertTrue(x_test_no.is_empty())

    # 2. Box fully in hidden zone (shadow)
    x = IntervalVector([[3.0, 4.0], [0.1, 0.4]])
    x_orig = IntervalVector(x)
    ctc_vis.contract(x)
    self.assertTrue(x.is_empty())

    x_test_no = IntervalVector(x_orig)
    ctc_nvis.contract(x_test_no)
    self.assertTrue(x_test_no == x_orig)

    # 3. Box behind the observer
    x = IntervalVector([[-2.0, -1.0], [-1.0, -0.2]])
    x_orig = IntervalVector(x)
    ctc_vis.contract(x)
    self.assertTrue(x == x_orig)

    x_test_no = IntervalVector(x_orig)
    ctc_nvis.contract(x_test_no)
    self.assertTrue(x_test_no.is_empty())

    # 4. Box on the side (outside the angular cone)
    x = IntervalVector([[1.0, 4.0], [2.5, 4.0]])
    x_orig = IntervalVector(x)
    ctc_vis.contract(x)
    self.assertTrue(x == x_orig)

    x_test_no = IntervalVector(x_orig)
    ctc_nvis.contract(x_test_no)
    self.assertTrue(x_test_no.is_empty())


if __name__ == '__main__':
  unittest.main()